unitycatalog-ai/integrations/autogen/src/unitycatalog/ai/autogen/toolkit.py

```diff
@@ -0,0 +1,162 @@
+import json
+from typing import Any, Callable, Dict, List, Optional
+
+from openai import pydantic_function_tool
+from packaging import version
+from pydantic import BaseModel, ConfigDict, Field, model_validator
+
+from autogen import ConversableAgent
+from autogen import __version__ as autogen_version
+from unitycatalog.ai.core.client import BaseFunctionClient
+from unitycatalog.ai.core.utils.client_utils import validate_or_set_default_client
+from unitycatalog.ai.core.utils.function_processing_utils import (
+    generate_function_input_params_schema,
+    get_tool_name,
+    process_function_names,
+)
+
+# Ensure the version of autogen is compatible
+if version.parse(autogen_version) >= version.parse("0.4.0"):
+    raise Exception(
+        "Autogen version should be less than 0.4.0 as the newer version has major API changes"
+    )
+
+
+class AutogenTool(BaseModel):
+    """
+    Model representing an Autogen tool.
+    """
+
+    fn: Callable = Field(
+        description="Callable that will be usedcto execute the UC Function and will be registered"
+    )
+
+    name: str = Field(
+        description="The name of the function.",
+    )
+    description: str = Field(
+        description="A brief description of the function's purpose.",
+    )
+    tool: Dict = Field(description="OpenAI compatible Tool Definition")
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Converts the Autogen Tool instance into a dictionary for the Autogen API.
+        """
+        return {"name": self.name, "description": self.description, "tool": self.tool}
+
+    def register_function(self, caller: ConversableAgent, executor: ConversableAgent) -> None:
+        """
+        Registers the function associated with the Autogen tool.
+
+        This method registers the callable function (`fn`) with the provided executor.
+        It also updates the tool signature with the caller to ensure that the function's definition
+        is correctly reflected in the overall toolset.
+
+        """
+        caller.update_tool_signature(self.tool, is_remove=False)
+        executor.register_function({self.name: executor._wrap_function(self.fn)})
```

Comment:
Is using a non-public API required here? Does autogen supply any public method for function registration apart from using `_wrap_function` ?

Reply:
Unfortunately yes.  all the public methods used require a function to be passed from which all the metadata is auto-inferred refer [here](https://github.com/microsoft/autogen/blob/v0.2.37/autogen/agentchat/conversable_agent.py#L2971) .In the current implementation we already get the metadata we need . but we need to use the  `_wrap_function` to make sure it is is logged in the trace and the output is in the required format